### PR TITLE
Generalize AR-Diffusion KV session capabilities

### PR DESCRIPTION
**Pull request by Jack47**
## Summary

- define a typed, model-neutral AR-Diffusion pipeline capability for KV geometry, worker-local KV branches, session lifetime, and optional warmup requests
- make `ARDiffusionModelRunner` own KV pool sizing, session reuse/reset/close, LRU eviction, failure cleanup, and state binding without importing or inspecting DreamZero internals
- generalize the KV runtime from fixed positive/negative storage to named `kv_branch` mappings and named, session-scoped cross-attention KV
- move DreamZero geometry, CFG mapping, model-state lifecycle, and warmup request construction into a DreamZero-side adapter
- document the contract and include a single-branch causal-DMD integration sketch

## Why

The experimental runner was coupled directly to DreamZero transformer fields, request schema, private state, and two fixed CFG branches. That prevented another autoregressive diffusion pipeline from opting into persistent session KV without changing the generic runtime.

The new contract keeps the public surface small and makes ownership explicit: the runner owns KV and session eviction; the pipeline owns model-specific non-KV state and only borrows a bound `ARDiffusionKVState` during one forward.

## Compatibility

- the default `DiffusionEngine` / `DiffusionModelRunner` routing is unchanged
- DreamZero keeps its existing `is_negative` model semantics and maps them to explicit `"positive"` / `"negative"` KV branches at the adapter boundary
- direct users of the experimental KV API have a mechanical migration documented in `docs/design/ar_diffusion_pipeline_capability.md`
- no LingBot model code or session/event/WebSocket infrastructure is added here

## Validation

- Spot CPU: `python3 -m pytest -q tests/ar_diffusion tests/dreamzero/test_pipeline_state.py` — **87 passed, 8 skipped**
- `ruff check ...` — passed
- `ruff format --check ...` — passed
- `git diff --check` — passed

Real DreamZero GPU eager/compile bitwise comparison is running separately and is intentionally not claimed in this draft.

## Review focus

- capability ownership and lifecycle semantics in `capability.py` and `runner.py`
- pool sizing and worker-local branch-slot mapping in `kv_cache/manager.py`
- single cleanup path for reset, close, LRU eviction, warmup, and forward exceptions
- DreamZero adapter equivalence, especially cross-attention KV population and internal window reset behavior

### CODE REVIEW

**Comment by david6666666** [comment]
@lishunyang12 @fhfuih  ptal thx

File: docs/design/ar_diffusion_pipeline_capability.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
+# AR-Diffusion pipeline capability
```
Comment (hsliuustc0106) [inline docs/design/ar_diffusion_pipeline_capability.md:1]: for the design doc, please refer to the issue https://github.com/vllm-project/vllm-omni/issues/5137 and related PR

File: vllm_omni/diffusion/models/dreamzero/pipeline_dreamzero.py [hunk 2]
```diff
@@ -92,18 +101,120 @@ class DreamZeroPipeline(nn.Module, CFGParallelMixin):
...
+    _NEGATIVE_BRANCH = "negative"
+    _ar_diffusion_kv_state = None
+
+    def ar_diffusion_kv_cache_spec(self) -> ARDiffusionKVCacheSpec:
+        """Describe DreamZero's local KV geometry to the generic runner."""
+        transformer = self.transformer
+        frame_tokens = int(transformer.frame_seqlen)
+        max_attention_tokens = int(transformer.blocks[0].self_attn.max_attention_size)
+        cfg_world = int(get_classifier_free_guidance_world_size())
+        negative_local_index = 0 if cfg_world >= 2 else 1
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/dreamzero/pipeline_dreamzero.py:121]: With CFG parallelism both logical branches share local index 0, but DreamZero eagerly populates both positive and negative cross-attention KV on every rank before branch dispatch. The negative write therefore overwrites the positive slot, so rank 0 can read negative conditioning for the positive branch. Populate only the rank-local branch or keep distinct cross-attention slots.

File: vllm_omni/experimental/ar_diffusion/kv_cache/manager.py [hunk 3]
```diff
@@ -260,30 +244,33 @@ def _cross_pool_bytes(length: int) -> int:
...
-        # (num_frame_per_block frame-blocks) for every branch THIS rank runs,
+        # (frames_per_block frame-blocks) for every KV branch THIS rank runs,
         # with a little eviction-transient headroom. The memory-fraction
         # heuristic can under-size this once block_size grows — e.g. frame-granular
         # paging at the true frame_seqlen makes each block larger and the pool
         # fewer-blocks — so guarantee the minimum the rollout cannot run without,
         # otherwise allocate_chunk hits an exhausted pool mid-forward.
-        min_blocks = self.local_branches * (config.window_chunks + self.num_frame_per_block) + 2
+        resident_blocks = config.sink_chunks + config.window_chunks
+        min_blocks = self.num_local_kv_branches * (resident_blocks + self.frames_per_block) + 2
```
Comment (hsliuustc0106) [inline vllm_omni/experimental/ar_diffusion/kv_cache/manager.py:265]: This floor only covers one resident session, while session_capacity sessions may retain full windows simultaneously. The pool can therefore exhaust before the count-based LRU evicts anything. Include session capacity in the required managed-block floor, or evict sessions on block pressure.

File: vllm_omni/experimental/ar_diffusion/runner.py [hunk 4]
```diff
@@ -55,355 +38,250 @@ def resolve_ar_diffusion_kv_config(od_config: OmniDiffusionConfig) -> ARDiffusio
...
-                "AR-Diffusion execute_model failed for session=%s; session state torn down "
-                "(pool blocks freed) — the next request starts a fresh session",
+                "AR-Diffusion forward failed for session=%s; KV and model state were released",
                 session_id,
             )
             raise
-        finally:
-            self.pipeline._ar_diffusion_kv_state = None
         if self.device is not None and torch.cuda.is_available():
             torch.accelerator.synchronize(self.device)
```
Comment (hsliuustc0106) [inline vllm_omni/experimental/ar_diffusion/runner.py]: This synchronization is outside the cleanup guard. An asynchronous kernel failure first reported here leaves the session and its pending managed blocks live, unlike a synchronous forward exception. Include synchronization in the guarded cleanup path.

File: vllm_omni/experimental/ar_diffusion/kv_cache/manager.py [hunk 5]
```diff
@@ -293,13 +280,13 @@ def _cross_pool_bytes(length: int) -> int:
         # Scratch blocks are outside KVCacheManager ownership. They hold current
         # denoise-step video KV when update_kv_cache=False and action/state KV in
         # both modes; they are reused every forward and never committed.
-        # Default 4: measured high-water is num_frame_per_block video blocks (2)
-        # + 1 action block per branch per forward, +1 margin.
-        scratch_per_branch = int(os.environ.get("AR_DIFFUSION_KV_SCRATCH_BLOCKS_PER_BRANCH", "4"))
-        if scratch_per_branch <= 0:
+        # Default 4 covers the current model high-water: several current video
+        # blocks plus optional model-specific scratch tokens.
+        scratch_per_kv_branch = int(os.environ.get("AR_DIFFUSION_KV_SCRATCH_BLOCKS_PER_BRANCH", "4"))
```
Comment (hsliuustc0106) [inline vllm_omni/experimental/ar_diffusion/kv_cache/manager.py]: The capability permits arbitrary frames_per_block, but non-committing video KV must fit in this fixed four-block scratch region. A valid pipeline with more than four frame blocks fails at runtime. Derive the minimum scratch capacity from the declared geometry instead of relying on an environment override.

File: vllm_omni/experimental/ar_diffusion/runner.py [hunk 6]
```diff
@@ -55,355 +38,250 @@ def resolve_ar_diffusion_kv_config(od_config: OmniDiffusionConfig) -> ARDiffusio
...
+            self._sessions[session_id] = state
+        self._sessions.move_to_end(session_id)
+        return state
+
+    @staticmethod
+    def _request_session(req: OmniDiffusionRequest) -> tuple[str, dict]:
+        extra_args = req.sampling_params.extra_args or {}
+        return str(extra_args.get("session_id") or "default"), extra_args
+
+    def execute_model(
```
Comment (hsliuustc0106) [inline vllm_omni/experimental/ar_diffusion/runner.py:237]: AR session state is bound only in request-mode execute_model. If step execution is enabled, or a future capable pipeline advertises request batching, DiffusionEngine routes through inherited stepwise or batch methods and bypasses this state. Reject those modes during initialization or implement equivalent binding for them.

**Comment by hsliuustc0106** [comment]
Non-blocking follow-up for the structured-config cutover: the current runtime still consumes legacy `StageConfig`/OmegaConf, so this PR can keep the existing path. In a future PR, please mirror `engine_backend` and `diffusion_model_runner_cls` into `_DiffusionConfigProjection`/`VllmOmniConfig`, model `ar_diffusion_kv_config` as typed configuration instead of a hidden `model_config` entry, use explicit optional sentinels for `sink_chunks`/`reset_at_boundary` precedence, validate AR-only constraints such as `max_num_seqs=1` during config construction, and add a DreamZero YAML-to-structured-config parity test.

**Comment by hsliuustc0106** [comment]
check #5244 as well for kv manager

**Comment by Jack47** [comment]
Updated after the final Local Zone validation.

The follow-up commits (`ee15ca4a`, `441ecc53`, `c735b199`, and `ecc18c62`) now:

- isolate cross-attention KV by logical branch and publish each branch transactionally;
- cap the current runner at one resident session, with reset/close/LRU/failure using the same release path;
- derive scratch capacity from declared pipeline geometry;
- reject step, batch, and multi-sequence execution before they can bypass AR state binding;
- document the current limits, #5137/#5244 relationship, and a single-branch model adapter.

Validation:

- targeted AR-Diffusion/DreamZero contracts on Spot L4: **107 passed**;
- real `GEAR-Dreams/DreamZero-DROID` checkpoint on a Local Zone Spot `p5e.48xlarge` H200;
- eager baseline versus candidate, two sequential same-session requests: video/actions are bitwise equal (`max_abs=0`);
- compile succeeds on both sides. Strict cross-process bitwise equality is not a valid gate for this backend: baseline is itself non-bitwise even when two runs reuse the same compiled cache. Baseline self-repeat video RMSE was `0.03094/0.04814`, candidate self-repeat was `0.03214/0.03713`, and baseline-to-candidate was `0.04554/0.04974`, with action differences also in the baseline variation range. No candidate-specific numerical regression was detected;
- DCO, Python 3.11/3.12 builds, pre-commit, and ReadTheDocs are green.

This replaces the earlier preliminary B200/shared-cache bitwise statement with the final controlled result.

**Comment by wjsuijlenh** [comment]
I studied the chunk-window eviction path closely for my own work, so one confirmation: freeing the gap between the sink and the recent window is the correct semantics, and the sink-aware visible window and the `sink_chunks`-counting pool floor are consistent with it. It also matches the dense implementation in #3701's `wan_fast.py`, which keeps the first `sink_tokens` in place and rolls the region behind them — two independent implementations arriving at the same behavior.

One coordination note, no change requested. My open PR #4487 touches the same DreamZero session code: it adds an opt-in path where the pipeline-owned session state lives in a small manager instead of the `_states` dict. Your `reset_ar_diffusion_session` / `close_ar_diffusion_session` hooks are exactly the notification surface that path needs, but they pop `_states` directly, so on the opt-in path a reset or eviction would not reach the state. Whichever PR lands second will route the two hooks through the same state-release call — flagging it now so the rebase brings no surprises.

**Review by hsliuustc0106 — APPROVED** [review]
LGTM

File: vllm_omni/experimental/ar_diffusion/runner.py [hunk 7]
```diff
@@ -56,359 +40,280 @@ def resolve_ar_diffusion_kv_config(od_config: OmniDiffusionConfig) -> ARDiffusio
...
-            num_frame_per_block,
-            local_attn_size,
-            chunk_size,
-            window_chunks,
-            window_chunks * chunk_size,
+        config = dataclasses.replace(
+            self.ar_diffusion_kv_config,
+            chunk_size=spec.tokens_per_frame,
+            window_chunks=self.ar_diffusion_kv_config.window_chunks or spec.window_frames,
+            sink_chunks=self.ar_diffusion_kv_config.sink_chunks or spec.sink_frames,
```
Comment (fhfuih) [inline vllm_omni/experimental/ar_diffusion/runner.py:120]: Overall looks good for now. One small question, why are there two different names `sink_chunks` and `sink_frames` ?

File: docs/design/ar_diffusion_pipeline_capability.md (hunk 1 above)
Comment (Jack47) [reply docs/design/ar_diffusion_pipeline_capability.md:1]: Addressed in `441ecc53`. The design note now links #5137 at the top, marks this contract experimental/non-normative, and separately documents how the current local AR session/KV ownership relates to #5244.

File: vllm_omni/diffusion/models/dreamzero/pipeline_dreamzero.py (hunk 2 above)
Comment (Jack47) [reply vllm_omni/diffusion/models/dreamzero/pipeline_dreamzero.py:121]: Addressed in `ee15ca4a`. Named cross-attention KV is now keyed by the logical KV branch, independently of `local_index`; `local_index` is used only for rank-local self/scratch storage. Population is also transactional. A focused test covers positive/negative branches sharing local slot 0 while retaining different cross-attention values.

File: vllm_omni/experimental/ar_diffusion/kv_cache/manager.py (hunk 3 above)
Comment (Jack47) [reply vllm_omni/experimental/ar_diffusion/kv_cache/manager.py:265]: Addressed in `441ecc53`. For the current FPS-first execution policy the runner caps effective residency at one session and passes that effective capacity to both managed and cross-attention pools. A different session ID evicts/releases the resident session before allocating the next one. The limitation and the requirements for future multi-session residency are documented.

File: vllm_omni/experimental/ar_diffusion/runner.py (hunk 4 above)
Comment (Jack47) [reply vllm_omni/experimental/ar_diffusion/runner.py]: Addressed in `441ecc53`. Device synchronization is now inside the guarded forward path, so an asynchronously surfaced kernel error uses the same `forward_exception` release path. The regression test verifies that the session and pending managed blocks are removed.

File: vllm_omni/experimental/ar_diffusion/kv_cache/manager.py (hunk 5 above)
Comment (Jack47) [reply vllm_omni/experimental/ar_diffusion/kv_cache/manager.py]: Addressed in `441ecc53`, with test geometry fixes in `c735b199`/`ecc18c62`. `ARDiffusionKVCacheSpec.max_scratch_tokens_per_branch` declares model-specific scratch demand; the pool derives its minimum from `frames_per_block + ceil(max_scratch_tokens_per_branch / tokens_per_frame)`. The environment override may only increase that minimum. DreamZero declares its action/state scratch tokens, and arbitrary geometry is covered by tests.

File: vllm_omni/experimental/ar_diffusion/runner.py (hunk 6 above)
Comment (Jack47) [reply vllm_omni/experimental/ar_diffusion/runner.py:237]: Addressed in `441ecc53`. Initialization now rejects `step_execution=True`, request-batch-capable pipelines, and `max_num_seqs > 1`. The inherited batch and step entry points also fail explicitly, so no execution route can silently bypass AR state binding. Focused tests cover the load-time failures.

File: vllm_omni/experimental/ar_diffusion/runner.py (hunk 7 above)
Comment (Jack47) [reply vllm_omni/experimental/ar_diffusion/runner.py:120]: `sink_frames` is the pipeline-facing semantic quantity: the number of leading latent frames retained as the attention sink. `sink_chunks` is the existing KV deployment/config term paired with `chunk_size` and `window_chunks`. In the current AR cache one chunk is one latent frame (`chunk_size = tokens_per_frame`), so the runner maps them directly. I kept the config name for backward compatibility; the distinction is now explained here, and a future typed-config cutover can consolidate the names without silently breaking existing YAML overrides.

**Comment by hsliuustc0106** [comment]
lgtm
